extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -301,4 +351,114 @@ private void removeCookie(RoutingContext context, TenantConfigContext configCont
             }
         }
     }
+
+    private boolean isLogout(RoutingContext context, TenantConfigContext configContext) {
+        Optional<String> logoutPath = configContext.oidcConfig.logout.path;
+
+        if (logoutPath.isPresent()) {
+            return context.request().absoluteURI().equals(
+                    buildUri(context, logoutPath.get()));
+        }
+
+        return false;
+    }
+
+    private SecurityIdentity trySilentRefresh(TenantConfigContext configContext, String idToken, String refreshToken,
+            RoutingContext context, IdentityProviderManager identityProviderManager) {
+        OidcTenantConfig config = configContext.oidcConfig;
+
+        if (config.token.refreshTimeout.isPresent()) {
+            OAuth2AuthProviderImpl auth = ((OAuth2AuthProviderImpl) configContext.auth);
+            JWT jwt = auth.getJWT();
+            JsonObject tokenJson;
+
+            try {
+                tokenJson = jwt.decode(idToken);
+            } catch (Exception cause) {
+                throw new AuthenticationCompletionException(cause);
+            }
+
+            Long iat = tokenJson.getLong("iat");
+            long now = System.currentTimeMillis() / 1000;
+            Integer timeout = config.token.refreshTimeout.get();
+
+            if (now - iat >= timeout) {
+                String sub = tokenJson.getString("sub");
+
+                TokenEntry entry = refreshingTokens.computeIfAbsent(sub,
+                        new Function<String, TokenEntry>() {
+                            @Override
+                            public TokenEntry apply(String token) {
+                                return new TokenEntry(refreshToken);
+                            }
+                        });
+
+                // synchronize on the entry associated with the current subject
+                synchronized (entry) {
+                    Uni<SecurityIdentity> cf = Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                        @Override
+                        public void accept(UniEmitter<? super SecurityIdentity> emitter) {
+                            OAuth2TokenImpl token = new OAuth2TokenImpl(configContext.auth, new JsonObject());
+
+                            // always get the last token
+                            token.principal().put("refresh_token", entry.getToken());
```

Comment:
Even if this is a new token it will still be refreshed?

Reply:
Yes. I don't think we need to control whether or not a refresh already happened but just get the latest.